analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java

```diff
@@ -246,9 +252,40 @@ private void addOrUpdateDayTruncatedActivity(UserAnonymizedEntityHolder userAnon
 		{
 			return Optional.empty();
 		}
-		return Optional.ofNullable(activityRepository.findOverlappingOfSameApp(dayActivity.get(), payload.deviceAnonymizedId,
+
+		List<Activity> overlappingOfSameApp = activityRepository.findOverlappingOfSameApp(dayActivity.get(), payload.deviceAnonymizedId,
 				matchingGoal.getActivityCategoryId(), payload.application.orElse(null), payload.startTime.toLocalDateTime(),
-				payload.endTime.toLocalDateTime()));
+				payload.endTime.toLocalDateTime());
+
+		// The prognosis is that there is no or one overlapping activity of the same app,
+		// because we don't expect the mobile app to post app activity that spans other existing same app activity
+		// (that indicates that there is something wrong in the app activity bookkeeping).
+		// Network activity (having payload.application == null) also is not expected to exist and overlap, as network activity
+		// has a very short time span.
+		// So log if there are multiple overlaps.
+
+		if (overlappingOfSameApp.isEmpty())
+		{
+			return Optional.empty();
+		}
+		else if (overlappingOfSameApp.size() == 1)
+		{
+			return Optional.of(overlappingOfSameApp.get(0));
+		}
+		else
+		{
+			String relatedTo = payload.application.isPresent()
+					? MessageFormat.format("same app ''{0}'' activity", payload.application.get()) : "network activity";
+			String overlappingActivities = overlappingOfSameApp.stream()
+					.map(overlappingActivity -> overlappingActivity.toString()).collect(Collectors.joining(", "));
+			logger.warn(
+					"Multiple overlapping activities of {0} found. The payload has start time {1} and end time {2}. The overlapping activities are: {3}. The day activity ID is {4} and the activity category ID is {5}.",
```

Comment:
I think the understandability would improve if the last part of the message was the collection of overlapping activities. That would require something like this:
```
"Multiple overlapping activities of {0} found. The payload has start time {1} and end time {2}. The day activity ID is {4} and the activity category ID is {5}. The overlapping activities are: {3}."
```

Note that SLF4J does not use sequence numbers for the arguments but expects the parameters to come in the right order, so the arguments need to be reordered and the numbers removed.

Reply:
Changed and fixed